src/acktest/adoption.py

```diff
@@ -0,0 +1,144 @@
+from .resources import random_suffix_name
+
+ADOPTED_RESOURCE_GROUP = "services.k8s.aws"
+ADOPTED_RESOURCE_VERSION = "v1alpha1"
+ADOPTED_RESOURCE_PLURAL = "adoptedresources"
+ADOPTED_RESOURCE_KIND = "AdoptedResource"
+
+ADOPTED_CONDITION_NAME = "ACK.Adopted"
+
+@dataclass(frozen=True)
+class AdoptedResourceAWSIdentifier:
+    """ Represents the base AWS identifier spec fields from the adopted resource CRD.
+
+    Additional keys need to be configured as allowlisted as part of the 
+    controller generator.
+    """
+    additionalKeys: Dict[str, str]
+
+@dataclass(frozen=True)
+class AdoptedResourceNameOrIDIdentifier(AdoptedResourceAWSIdentifier):
+    nameOrID: str
+
+@dataclass(frozen=True)
+class AdoptedResourceARNIdentifier(AdoptedResourceAWSIdentifier):
+    arn: str
+
+@dataclass(frozen=True)
+class AdoptedResourceKubernetesIdentifiers:
+    """ Represents the Kubernetes spec fields from the adopted resource CRD.
+
+    This class purposefully does not expose metadata configuration elements, as
+    those are configured by the test fixture.
+    """
+    group: str
+    kind: str
+
+@dataclass(frozen=True)
+class AdoptedResourceSpec:
+    """ Represents the adopted resource CRD spec fields.
+    """
+    aws: AdoptedResourceAWSIdentifier
+    kubernetes: AdoptedResourceKubernetesIdentifiers
+
+class AbstractAdoptionTest(ABC):
+    RESOURCE_PLURAL: str = ""
+    RESOURCE_VERSION: str = ""
+    TARGET_NAMESPACE: str = "default"
+
+    _spec: AdoptedResourceSpec  = None
+    _reference: k8s.CustomResourceReference
+
+    @abstractmethod
+    def bootstrap_resource(self):
+        pass
+
+    @abstractmethod
+    def cleanup_resource(self):
+        pass
+
+    @abstractmethod
+    def get_resource_spec(self) -> AdoptedResourceSpec:
+        pass
+
+    def _generate_resource_name(self) -> str:
+        return (random_suffix_name(f"adopted-{self._spec.kubernetes.kind}", 32, delimiter="-")).lower()
+
+    def _create_adopted_resource(self, resource_name: str) -> Tuple[k8s.CustomResourceReference, Dict]:
+        spec_dict = asdict(self._spec)
+        body_dict = {
+            "apiVersion": f"{ADOPTED_RESOURCE_GROUP}/{ADOPTED_RESOURCE_VERSION}",
+            "kind": ADOPTED_RESOURCE_KIND,
+            "metadata": {
+                "name": resource_name,
+                "namespace": self.TARGET_NAMESPACE
+            },
+            "spec": spec_dict
+        }
+
+        reference = k8s.CustomResourceReference(ADOPTED_RESOURCE_GROUP, ADOPTED_RESOURCE_VERSION,
+            ADOPTED_RESOURCE_PLURAL, resource_name, namespace=self.TARGET_NAMESPACE)
+
+        return (reference, k8s.create_custom_resource(reference, body_dict))
+        
+
+    def _assert_adoption_status(self):
+        cr = k8s.wait_resource_consumed_by_controller(self._reference)
+
+        assert cr is not None
+        assert k8s.get_resource_exists(self._reference)
+        assert k8s.wait_on_condition(self._reference, ADOPTED_CONDITION_NAME, "True")
```

Comment:
consider using the newly-added [`condition.assert_type_status`](https://github.com/aws-controllers-k8s/test-infra/blob/75752b2d9cd67aad898bf281c90f1d8b774391a5/src/acktest/k8s/condition.py#L28-L65) here:

```python
from .k8s import condition
...
condition.assert_type_status(self._reference, condition.CONDITION_TYPE_ADOPTED, True)
```

Reply:
If you really want to wait with a timeout, I can add new waiter functions to that module...